StyleCop.Analyzers/StyleCop.Analyzers.Test/MaintainabilityRules/SA1406UnitTests.cs

```diff
@@ -0,0 +1,158 @@
+        {
+            var testCode = @"using System.Diagnostics;
+public class Foo
+{
+    public void Bar()
+    {
+        Debug.Fail(""     "");
+    }
+}";
+
+            var expected = new[]
+            {
+                new DiagnosticResult
+                {
+                    Id = DiagnosticId,
+                    Message = "Debug.Fail must provide message text",
+                    Severity = DiagnosticSeverity.Warning,
+                    Locations =
+                        new[]
+                        {
+                            new DiagnosticResultLocation("Test0.cs", 6, 9)
+                        }
+                }
+            };
+
+            await VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestNullMessage()
+        {
+            var testCode = @"using System.Diagnostics;
+public class Foo
+{
+    public void Bar()
+    {
+        Debug.Fail(null);
+    }
+}";
+
+            var expected = new[]
+            {
+                new DiagnosticResult
+                {
+                    Id = DiagnosticId,
+                    Message = "Debug.Fail must provide message text",
+                    Severity = DiagnosticSeverity.Warning,
+                    Locations =
+                        new[]
+                        {
+                            new DiagnosticResultLocation("Test0.cs", 6, 9)
+                        }
+                }
+            };
+
+            await VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestConstantMessage()
+        {
+            var testCode = @"using System.Diagnostics;
+public class Foo
+{
+    public void Bar()
+    {
+        Debug.Fail(""A message"");
+    }
+}";
+
+            await VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestNotConstantMessage()
+        {
+            var testCode = @"using System.Diagnostics;
+public class Foo
+{
+    public void Bar()
+    {
+        string message = ""A message"";
+        Debug.Fail(message);
+    }
+}";
+
+            await VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestWrongMethod()
+        {
+            var testCode = @"using System.Diagnostics;
+public class Foo
+{
+    public void Bar()
+    {
+        Debug.Write(null);
+    }
+}";
+
+            await VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        [TestMethod]
+        public async Task TestWrongDebugClass()
+        {
+            var testCode = @"
+public class Foo
+{
+    public void Bar()
+    {
+        string message = ""A message"";
+        Debug.Fail(message);
```

Comment:
Nice catch to think of this test. :+1: However, the particular input you chose would not produce a diagnostic message even if it was analyzed as the normal `System.Diagnostics.Debug.Fail` method. You should add a test that would normally produce a diagnostic, but does not here only because the `Debug` class is not the expected one.